### PR TITLE
Upgrade hydro power representation to use energy budget and dispatch capacity factors 

### PR DESCRIPTION
**Pull request by sriharisundar**
### Is this a formatting PR (does not change outputs) or a substantive PR (changes outputs)?
Substantive PR

### High-level description
Use the hydro capacity factor values to model dispatchableHD plants as generator storages and nondispatchableHD plants as generators with variable capacities.

### Implementation notes
1. Added required hydro files for the test case
2. Did not add tests for new capabilities, but tested that the code runs for the test system.

### Which ReEDS-2.0 version tag is this PR based on?
[v2024.2.0](https://github.com/NREL/ReEDS-2.0/releases/tag/v2024.2.0)

### Known incompatibilities (if any)
Since it uses the latest monthly implementation of hydro capacity factors, using a prior iteration of ReEDS which uses seasonal hydro capacity factors will fail.

#### Use test/test_benchmark.jl and follow instructions there

Using 10 samples and 1 core on M2 MacBook Pro

|Case | R2P runtime | PRAS runtime | PRAS outputs |
| -------- | ------- | ------- | ------- |
| Before    | 30.912s | 178.44s | LOLE=1518.2, nEUE=5.29 |
| Before    | 39.112s | 399.548s | LOLE=1577.3, nEUE=5.68 |

### CODE REVIEW

**Comment by sriharisundar** [comment]
@patrickbrown4 We have functionality now to fallback to the old method for processing HD generators. R2P related code (and cases.csv etc) in ReEDS needs to be modified to support it. For example, [the R2P call ](https://github.nrel.gov/ReEDS/ReEDS-2.0/blob/cde699a8de90a9edc26ec4815109080f387bcfd5/ReEDS_Augur/run_pras.jl#L276) needs to be 
```
pras_system = ReEDS2PRAS.reeds_to_pras(
    args["reedscase"],
    args["solve_year"],
    args["timesteps"],
    args["weather_year"];
    proc_hd_as_convcap=args["proc_hd_as_convcap"]
)
```
where  `args["proc_hd_as_convcap"]` needs to be `true` if we want the old method and `false` if we want the updated methods (default in R2P).

File: src/main/parse_reeds_data.jl [hunk 1]
```diff
@@ -102,8 +105,26 @@ function parse_reeds_data(
...
+        hydro_disp_gens,
+        hydro_non_disp_gens,
+        forced_outage_dict,
+        ReEDS_data,
+        year,
+        WEATHERYEAR,
+        timesteps,
+        #        min_year,
+        user_inputs;
+        proc_hd_as_convcap=proc_hd_as_convcap,
```
Comment (patrickbrown4) [inline src/main/parse_reeds_data.jl]: Thanks for adding this!
We don't abbreviate hydro as 'hd' anywhere else and we try not to use "conventional", do how about changing it to something like `hydro_no_energylim`?

File: src/models/Variable_Gen.jl [hunk 2]
```diff
@@ -73,7 +78,10 @@ struct Variable_Gen <: Generator
         MTTR = 24,
     )
         all(0.0 .<= capacity .<= installed_capacity) ||
-            error("$(name) time series has values < 0 or > installed capacity
+        # TODO: This is a warning because some hydro capacity factors are > 1.0
+        #       giving an error here, but because of hydro budgets that is physically
+        #       ok, considering historical dispatches
+            @warn ("$(name) time series has values < 0 or > installed capacity
```
Comment (patrickbrown4) [inline src/models/Variable_Gen.jl]: could you add an `if` statement that bypasses the warning for hydro techs?

File: test/testutils.jl [hunk 3]
```diff
@@ -119,6 +119,7 @@ function compare_line_capacities(pras_system::PRAS.SystemModel, ReEDSfilepath, y
         end
         mw_out_fwd, mw_out_bck = get_pras_line_capacity(pras_system, Int.(retained_lines))
 
+        # TODO: Maybe some problems with using floor, round, and neither in different cases
```
Comment (patrickbrown4) [inline test/testutils.jl:122]: If you think we should look into this, could you open an issue and delete the TODO? (For ReEDS we've tried to remove TODO's from the code since they're hard to keep track of, so it'd be nice to do the same here.)

Could you do the same for the other TODO's you've added above?

File: src/utils/reeds_input_parsing.jl [hunk 4]
```diff
@@ -321,3 +375,39 @@ function get_storage_energy_capacity_data(data::ReEDSdatapaths)
...
+            ["JUN" 720 "summer"]
+            ["JUL" 744 "summer"]
+            ["AUG" 744 "summer"]
+            ["SEP" 720 "fall"]
+            ["OCT" 744 "fall"]
+            ["NOV" 720 "winter"]
+            ["DEC" 744 "winter"]
+        ],
+        Vector(["month", "numhrs", "season"]),
+    )
```
Comment (patrickbrown4) [inline src/utils/reeds_input_parsing.jl]: For leap years we drop Dec 31 instead of Feb 29 (I think that decision is made way upstream in SAM, or if not then in reV). Could you read this table from ReEDS instead of hard-coding it? You can get it from inputs_case/hmap_7yr.csv

File: src/models/Variable_Gen.jl [hunk 5]
```diff
@@ -1,16 +1,21 @@
...
+    "dupv",
+    "upv",
+    "distpv",
+    "csp",
+    "hydnd",
+    "hydsnd",
+    "hydund",
+    "hydnpnd",
+    "hydend",
+]
```
Comment (patrickbrown4) [inline src/models/Variable_Gen.jl:13]: It looks like you could get this from the VRE, CSP, and HYDRO_ND columns in inputs_case/tech-subset-table.csv. 

An alternative would be to add a "PRAS_VRE" (or something like that) column to tech-subset-table.csv and read it directly from there. I think I like that better because then all the sets are defined in one place, but if we do that we'd need to do a parallel ReEDS PR.

File: src/utils/runchecks.jl [hunk 6]
```diff
@@ -197,4 +197,37 @@ function run_checks(data::ReEDSdatapaths)
...
+    io_cfadj, bool_cfadj = check_file(filepath_cfadj)
+
+    if (bool)
+        close(io)
+    elseif (bool)
+        close(io_cfadj)
+    else
+        error(
+            "Hydro capacity adjustment data is not available in ReEDS results. You are either using a ReEDS version not compatible with ReEDS2PRAS (or) the ReEDS case results location passed is erroneous (or) you don't have access to the file/ deleted it.",
+        )
```
Comment (patrickbrown4) [inline src/utils/runchecks.jl]: Can you mention the full filepath ({ReEDScase}/inputs_case/hydcfadj.csv) in the error? Same for the above error message.

File: src/utils/reeds_input_parsing.jl [hunk 7]
```diff
@@ -293,6 +293,60 @@ function get_ICAP_data(data::ReEDSdatapaths)
...
+# TODO: Include test for this function
+function get_hydro_data(data::ReEDSdatapaths)
+    filepath_cf = joinpath(data.ReEDSfilepath, "inputs_case", "hydcf.csv")
+    hydcf = DataFrames.DataFrame(CSV.File(filepath_cf))
+
+    # Rename plant types as techtypes and capacity are lowercase
+    DataFrames.rename!(hydcf, [:"*i"] .=> [:i])
+    hydcf.i = lowercase.(hydcf.i)
+    # Subset to required run year, unclear if subset has to be 
+    # the ReEDS target year or the weather data years
```
Comment (patrickbrown4) [inline src/utils/reeds_input_parsing.jl]: Can "unclear if subset has to be the ReEDS target year or the weather data years" be deleted or is there still an issue here?

File: src/utils/reeds_input_parsing.jl [hunk 8]
```diff
@@ -293,6 +293,60 @@ function get_ICAP_data(data::ReEDSdatapaths)
...
+    hydcf = filter(x -> x.t == data.year, hydcf)
+
+    filepath_capadj = joinpath(data.ReEDSfilepath, "inputs_case", "hydcapadj.csv")
+
+    # TODO: Remove after PR1098 merged on ReEDS-2.0
+    #       and future ReEDS runs use that version
+    if !(isfile(filepath_capadj))
+        filepath_capadj = joinpath(data.ReEDSfilepath, "inputs_case", "hydcfadj.csv")
+    end
```
Comment (patrickbrown4) [inline src/utils/reeds_input_parsing.jl]: This PR has been merged so this comment (and the reassignment if I understand correctly) should be deleted

File: src/utils/reeds_data_parsing.jl [hunk 9]
```diff
@@ -363,6 +387,219 @@ function process_vg(
...
+    user_inputs::Dict{Any, Any};
+    proc_hd_as_convcap=false,
+    unitsize_dict=nothing
+)
+
+    if proc_hd_as_convcap && !isnothing(unitsize_dict)
+
+        @info "Processing HD generators as thermal generators..."
+
+        process_thermals_with_disaggregation(
```
Comment (patrickbrown4) [inline src/utils/reeds_data_parsing.jl]: What's actually happening here? Should something be reassigned to the output of these functions? Or do they modify global variables (which is scary to me)?

File: src/utils/reeds_data_parsing.jl [hunk 10]
```diff
@@ -363,6 +387,219 @@ function process_vg(
...
+        name = "$(category)_$(string(row.r))"
+        region = string(row.r)
+
+        if category in keys(FOR_dict)
+            gen_for = FOR_dict[category]
+        else
+            gen_for = 0.00 # make this 0 for vg if no match
+        end
+
+        # - Charging to genstore is limited by charge_capacity whether from grid or from
```
Comment (patrickbrown4) [inline src/utils/reeds_data_parsing.jl:602]: Hydro shouldn't be able to charge from the grid; it should only accept inflows. I'm still learning julia so don't follow all the steps but are you allowing grid charging? If not, maybe modify this comment to clarify that you only allow charging via inflows, not from the grid.

File: src/utils/reeds_data_parsing.jl [hunk 11]
```diff
@@ -363,6 +387,219 @@ function process_vg(
...
+                region_name = region,
+                charge_cap = repeat(monthly_energy, num_years),
+                discharge_cap = repeat(dispatch_limit, num_years),
+                energy_cap = repeat(energy_cap, num_years),
+                inflow = repeat(monthly_energy, num_years),
+                grid_withdrawl_cap = zeros(timesteps),
+                grid_inj_cap = repeat(dispatch_limit, num_years),
+                type = category,
+                legacy = "New",
+                FOR = 0.0, #TODO: update FORs. For now assuming no outages.
```
Comment (patrickbrown4) [inline src/utils/reeds_data_parsing.jl:623]: Looks like this still needs to be addressed. Hydro outage rates are copied from ReEDS-2.0/inputs/plant_characteristics/outage_forced.csv to {ReEDScase}/inputs_case/outage_forced.csv

**Review by patrickbrown4 — CHANGES_REQUESTED** [review]
Thanks for doing this! I left a few comments; the main one is that I'd like to get TODO's out of the code and into issues if they still need to be addressed (that's our standard in ReEDS and I think it helps keep track of issues better than notes in the code).

I'm still learning julia so it would probably be helpful for @scdhulipala to take a quick look too.

File: src/utils/reeds_data_parsing.jl (hunk 10 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:602]: We set grid_withdrawl_cap to be zero which means it cannot charge from the grid, only from inflows. But the unit can charge only if it is allowed to charge according to charge_cap.

File: src/utils/reeds_data_parsing.jl (hunk 9 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl]: Good catch. The `generators_array` variable is being modified within those function by https://github.com/NREL/ReEDS2PRAS/blob/b1a7b8817ebffef3ee3bcafd8b07ed9d914669f2/src/utils/reeds_data_parsing.jl#L303-L304.

File: src/utils/runchecks.jl (hunk 6 above)
Comment (sriharisundar) [reply src/utils/runchecks.jl]: Just doing the filename as that seems to be the convention followed above.

File: test/testutils.jl (hunk 3 above)
Comment (sriharisundar) [reply test/testutils.jl:122]: These tests are from before, and I don't think we run these at all. We do need to discuss writing better tests.

File: src/models/Variable_Gen.jl (hunk 5 above)
Comment (sriharisundar) [reply src/models/Variable_Gen.jl:13]: Can we do this in a separate PR?

File: src/models/Variable_Gen.jl (hunk 2 above)
Comment (sriharisundar) [reply src/models/Variable_Gen.jl]: Done

Comment (sriharisundar) [reply src/models/Variable_Gen.jl]: Will throw an error if not hydro and capacity somewhere is greater than installed capacity

File: src/utils/reeds_input_parsing.jl (hunk 7 above)
Comment (sriharisundar) [reply src/utils/reeds_input_parsing.jl]: I think it is not an issue for now. We just get the same solve year data for all provided weather years (for example if solve year is 2050, we get 2050 data for 2007-2013)

File: src/utils/reeds_data_parsing.jl (hunk 11 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:623]: As discussed, I am punting this for later as we do not disaggregate to individual hydro units here and applying FORs might not be appropriate at the regional level.

File: src/models/Variable_Gen.jl (hunk 5 above)
Comment (patrickbrown4) [reply src/models/Variable_Gen.jl:13]: Yeah I can take care of this on the ReEDS repo side after you're done with this PR

File: src/utils/reeds_input_parsing.jl (hunk 4 above)
Comment (sriharisundar) [reply src/utils/reeds_input_parsing.jl]: I am inclined to preserve this here so as to stay consistent with [[ReEDS-2.0_Input_Processing](https://github.nrel.gov/ReEDS/ReEDS-2.0_Input_Processing)](https://github.nrel.gov/ReEDS/ReEDS-2.0_Input_Processing/blob/3588752c2e7ddb827e60685eb52c0ad5651b68ea/hydro-capacityfactors/capacityfactor.py#L4)
Also, I checked the version of the hmap_7yr.csv file I have, and it seems to not have the seasons for all years. Maybe I am looking at the wrong file.

File: test/testutils.jl (hunk 3 above)
Comment (patrickbrown4) [reply test/testutils.jl:122]: Ok, since we're moving all of this to the ReEDS repo anyway, we can change the TODOs into github issues at that time.

File: src/utils/reeds_input_parsing.jl [hunk 12]
```diff
@@ -293,6 +293,54 @@ function get_ICAP_data(data::ReEDSdatapaths)
...
+# TODO: Include test for this function
+function get_hydro_data(data::ReEDSdatapaths)
+    filepath_cf = joinpath(data.ReEDSfilepath, "inputs_case", "hydcf.csv")
+    hydcf = DataFrames.DataFrame(CSV.File(filepath_cf))
+
+    # Rename plant types as techtypes and capacity are lowercase
+    DataFrames.rename!(hydcf, [:"*i"] .=> [:i])
+    hydcf.i = lowercase.(hydcf.i)
+    # Subset to required run year, unclear if subset has to be 
+    # the ReEDS target year or the weather data years
```
Comment (patrickbrown4) [inline src/utils/reeds_input_parsing.jl]:
```suggestion
    # Subset to ReEDS model year
```

File: src/utils/reeds_data_parsing.jl [hunk 13]
```diff
@@ -25,19 +25,9 @@ function process_regions_and_load(ReEDS_data, weather_year::Int, timesteps::Int)
...
-        load_year = load_data[:, indices[1:8760]]
-        load_timesteps = repeat(load_year, 1, floor(Int, timesteps / 8760))
-        if timesteps % 8760 > 0
-            load_timesteps =
-                hcat(load_timesteps, load_data[:, indices[1:(timesteps % 8760)]])
-        end
-    else
-        load_timesteps = load_data[:, indices[1:timesteps]]
-    end
+    indices = findfirst(i -> (i == (slicer - 1)), load_info["axis1_label1"])
```
Comment (patrickbrown4) [inline src/utils/reeds_data_parsing.jl]: The version you added in the ReEDS repo uses `findall` instead of `findfirst` - does the difference matter?

https://github.nrel.gov/ReEDS/ReEDS-2.0/pull/1405/files

File: Project.toml [hunk 14]
```diff
@@ -1,6 +1,6 @@
 name = "ReEDS2PRAS"
 uuid = "c1db39dc-6a9d-11ed-0c9a-05c3ea2a1475"
-authors = ["Luke Lavin <luke.lavin@nrel.gov>", "Surya Chandan Dhulipala <suryachandan.dhulipala@nrel.gov>", "Brandon Norton Benton <brandon.benton@nrel.gov"]
+authors = ["Luke Lavin <luke.lavin@nrel.gov>", "Surya Chandan Dhulipala <suryachandan.dhulipala@nrel.gov>", "Brandon Norton Benton <brandon.benton@nrel.gov>", "Hari Sundar <hari.sundar@nrel.gov"]
 version = "0.3.0"
```
Comment (scdhulipala) [inline Project.toml:4]: We should probably tag a new version once we get this merged and update the compatability table.

File: src/main/create_pras_system.jl [hunk 15]
```diff
@@ -124,8 +124,7 @@ function create_pras_system(
     gen_stor_names =
         isempty(get_name.(sorted_gen_stors)) ? String[] : get_name.(sorted_gen_stors)
     gen_stor_cats =
-        isempty(get_category.(sorted_gen_stors)) ? String[] :
-        get_category.(sorted_gen_stors)
+        isempty(get_type.(sorted_gen_stors)) ? String[] : get_type.(sorted_gen_stors)
```
Comment (scdhulipala) [inline src/main/create_pras_system.jl:127]: Any idea why this was changed to type from category? We defined get_category to include both legacy and type.

File: src/utils/reeds_data_parsing.jl [hunk 16]
```diff
@@ -363,6 +445,218 @@ function process_vg(
...
+                    (hydcapadj.i .== row.i) .&& (hydcapadj.r .== row.r),
+                    [:month, :value],
+                ]
+                dispatch_limit[reqd_slice] .=
+                    (filter(x -> (x.month == mon_row.month), capacity_adjust)[
+                        :,
+                        :value,
+                    ] * row.MW_sum)[1]
+            catch e
+                @error "$(row.r),$(row.i),$(e)"
```
Comment (patrickbrown4) [inline src/utils/reeds_data_parsing.jl]: Should we get rid of this try/catch and let it fail if there's an error? I doubt people will check every log file carefully enough to notice it.

File: src/utils/reeds_data_parsing.jl (hunk 11 above)
Comment (patrickbrown4) [reply src/utils/reeds_data_parsing.jl:623]: Do we uniformly derate the hydro capacity by the FOR when using the new method? (Did we do that in the old method? Sorry, I'm not sure.) The FOR depends on the hydro type but linearly derating seems most consistent with the treatment of other non-disaggregated techs.
https://github.nrel.gov/ReEDS/ReEDS-2.0/blob/06e282a3011341914f89176e37a09d19d09b4e00/inputs/plant_characteristics/outage_forced.csv#L40-L50

**Review by patrickbrown4 — APPROVED** [review]
I'll go through it again when we add it to the ReEDS repo, but otherwise going ahead and approving aside from a few small comments. Thanks!

File: src/utils/reeds_data_parsing.jl (hunk 13 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl]: I'll update it to what's on the ReEDS repo, but this works too.

File: src/main/parse_reeds_data.jl [hunk 17]
```diff
@@ -48,13 +49,14 @@ function parse_reeds_data(
 
     # Create Generator Objects
     # **TODO: Should 0 MW generators be allowed after disaggregation?
-    # **TODO: Should hydro be split out as a generator-storage?
     # **TODO: is it important to also handle planned outages?
     @info(
-        "splitting thermal, storage, vg generator types from installed " *
+        "splitting thermal, storage, vg generator, hydro generator types from installed " *
```
Comment (scdhulipala) [inline src/main/parse_reeds_data.jl]: Can we change this to : "splitting thermal, storage, variable and hydro generator types from installed ReEDS capacities..."

File: src/reeds_to_pras.jl [hunk 18]
```diff
@@ -49,13 +55,15 @@ function reeds_to_pras(
...
     @info "Parsing ReEDS data and creating ReEDS2PRAS objects..."
+    # TODO: Update min year variable
     out = parse_reeds_data(
         ReEDS_data_filepaths,
         weather_year,
         timesteps,
         solve_year,
         2007,
-        user_inputs,
+        user_inputs;
```
Comment (scdhulipala) [inline src/reeds_to_pras.jl]: You don't need a ; after user_inputs because we are calling _parse_reeds_data_ here and not defining.

File: src/utils/reeds_data_parsing.jl [hunk 19]
```diff
@@ -244,7 +256,8 @@ function process_thermals_with_disaggregation(
     unitsize_dict::Dict,
     timesteps::Int,
     year::Int,
-    user_inputs::Dict{Any, Any},
+    user_inputs::Dict{Any, Any};
+    all_generators = Generator[],
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl:260]: I'm not sure why we need this here?

File: src/utils/reeds_data_parsing.jl [hunk 20]
```diff
@@ -283,6 +295,74 @@ function process_thermals_with_disaggregation(
     return all_generators
 end
 
+"""
+    We use this function if we want to proces hydroelectric generators as fixed capacities.
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl]: process

File: src/utils/reeds_data_parsing.jl [hunk 21]
```diff
@@ -283,6 +295,74 @@ function process_thermals_with_disaggregation(
...
+        Dictionary of Forced Outage Rates (FORs) for each resource
+    timesteps : int
+        Number of slices to disaggregate the capacity
+    year : int
+        Year associated with the capacity
+
+    Returns
+    -------
+"""
+function process_hd_as_generator!(
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl:323]: Why can't we re-use _process_thermal_with_disaggregation_ to avoid code duplication? If you do'nt want to handle this in this PR, we can do it in a different PR.

File: src/utils/reeds_data_parsing.jl [hunk 22]
```diff
@@ -363,6 +445,218 @@ function process_vg(
     return generators_array
 end
 
+"""
+    Parameters
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl:449]: Update this to reflect the correct format of the _process_hydro_ function

File: src/utils/reeds_data_parsing.jl [hunk 23]
```diff
@@ -325,6 +405,8 @@ function process_vg(
     cf_info = get_vg_cf_data(ReEDS_data)
 
     vg_profiles = cf_info["block0_values"]
+    # TODO: (SSH) This logic will fail if timesteps != 8760
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl:408]: Can you open an issue? I feel like min_year will be 2007 in the near future because ReEDS_Augur writes out this file.

File: src/utils/reeds_data_parsing.jl [hunk 24]
```diff
@@ -363,6 +445,218 @@ function process_vg(
...
+    year::Int64,
+    weather_year::Int,
+    timesteps::Int,
+    user_inputs::Dict{Any, Any};
+    hydro_energylim = false,
+    unitsize_dict = nothing,
+)
+
+    # If we do not impose energy limits on hydro and model it as fixed capacity
+    if !(hydro_energylim) && !isnothing(unitsize_dict)
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl]: Why would unitsize_dict be nothing? The process will fail if we can't parse the ATB capacity CSV, right?

Comment (scdhulipala) [reply src/utils/reeds_data_parsing.jl]: I think you should just check for _hydro_energy_lim_.

File: src/utils/reeds_data_parsing.jl [hunk 25]
```diff
@@ -363,6 +445,218 @@ function process_vg(
...
+    hydro_disp_capacities::DataFrames.DataFrame,
+    hydro_non_disp_capacities::DataFrames.DataFrame,
+    FOR_dict::Dict,
+    ReEDS_data,
+    year::Int64,
+    weather_year::Int,
+    timesteps::Int,
+    user_inputs::Dict{Any, Any};
+    hydro_energylim = false,
+    unitsize_dict = nothing,
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl]: Why is this an optional arg?

File: src/utils/reeds_data_parsing.jl [hunk 26]
```diff
@@ -363,6 +445,218 @@ function process_vg(
...
+            timesteps,
+            year,
+            user_inputs;
+        )
+        genstor_array = Gen_Storage[]
+
+        return generators_array, genstor_array
+    end
+
+    hydcf, hydcapadj = get_hydro_data(ReEDS_data)
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl:532]: I think this should block should be else block of the above if condition?

File: src/utils/reeds_input_parsing.jl [hunk 27]
```diff
@@ -321,3 +368,39 @@ function get_storage_energy_capacity_data(data::ReEDSdatapaths)
...
+
+    Parameters
+    ----------
+    None
+
+    Returns
+    -------
+    monhours: DataFrame
+"""
+function monhours()
```
Comment (scdhulipala) [inline src/utils/reeds_input_parsing.jl:420]: What do you think about this implementation?

monthnum_to_season_mapping = Dict(
    [1,2,11,12] => "winter",
    [3,4,5] => "spring",
    [6,7,8] => "summer",
    [9,10] => "fall"
)

function cumsum!(collection::Vector{monthhour})
    sum = 0
    for element in collection
        sum = sum+element.numhrs
        element.cumsum = sum
    end
end

mutable struct monthhour
    month::String
    numhrs::Int64
    season::String
    cumsum::Int64
    
    # Inner Constructors & Checks
    function monthhour(
        month = "month",
        numhrs = 10,
        season = "season";
        cumsum = 0
    )
        return new(month,numhrs,season, cumsum)
    end
end
monthours = monthhour[]
start_date = Dates.Date("2021-01", "yyyy-mm")
for i in range(0,length=12)
    new_date = start_date + Dates.Month(i)
    season = collect(values(monthnum_to_season_mapping))[findfirst(in.(i+1,keys(monthnum_to_season_mapping)))]
    push!(monthours, monthhour(uppercase(Dates.monthabbr(i+1)), Dates.daysinmonth(new_date)*24, season))
end

cumsum!(monthours)

File: src/utils/reeds_data_parsing.jl [hunk 28]
```diff
@@ -363,6 +445,218 @@ function process_vg(
...
+        reg_plant_subset =
+            filter(x -> (x.r == row.r && x.i == row.i), hydcf)[:, [:month, :value]]
+        monthly_energy = zeros(timesteps_year)
+        dispatch_limit = zeros(timesteps_year)
+        energy_cap = zeros(timesteps_year)
+        for (idx_mon, mon_row) in enumerate(DataFrames.eachrow(monthhours))
+            #@debug reg_plant_subset[findall(x->startswith(mon_row.season,x),reg_plant_subset.szn),:value]
+            try
+                reqd_slice = (mon_row.cumhrs - mon_row.numhrs + 1):(mon_row.cumhrs)
+                monthly_energy[first(reqd_slice)] = (mon_row.numhrs * filter(
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl]: Can we save (mon_row.numhrs * filter(
                    x -> (x.month == mon_row.month),
                    reg_plant_subset,
                )
as a variable so we don't repeat it again?

File: src/utils/reeds_data_parsing.jl [hunk 29]
```diff
@@ -363,6 +445,218 @@ function process_vg(
...
+
+                energy_cap[reqd_slice] .= (mon_row.numhrs * filter(
+                    x -> (x.month == mon_row.month),
+                    reg_plant_subset,
+                )[
+                    :,
+                    :value,
+                ] * row.MW_sum)[1]
+
+                capacity_adjust = hydcapadj[
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl:571]: This can be out of the _monthhour_ loop, right?

File: src/utils/reeds_data_parsing.jl (hunk 16 above)
Comment (scdhulipala) [reply src/utils/reeds_data_parsing.jl]: I second this.

File: src/utils/reeds_data_parsing.jl [hunk 30]
```diff
@@ -363,6 +445,218 @@ function process_vg(
...
+                region_name = region,
+                charge_cap = repeat(monthly_energy, num_years),
+                discharge_cap = repeat(dispatch_limit, num_years),
+                energy_cap = repeat(energy_cap, num_years),
+                inflow = repeat(monthly_energy, num_years),
+                grid_withdrawl_cap = zeros(timesteps),
+                grid_inj_cap = repeat(dispatch_limit, num_years),
+                type = category,
+                legacy = "New",
+                FOR = 0.0, #TODO: update FORs. For now assuming no outages.
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl:623]: @sriharisundar  Just a clarification here, once we disaggreagte Hydro generators, the plan is to derate energy_cap, right?

File: src/utils/reeds_data_parsing.jl [hunk 31]
```diff
@@ -363,6 +445,218 @@ function process_vg(
...
+            generators_array,
+            Variable_Gen(
+                name = name,
+                timesteps = timesteps,
+                region_name = region,
+                installed_capacity = row.MW_sum,
+                capacity = repeat(hourly_capacity, num_years),
+                type = category,
+                legacy = "New",
+                FOR = 0.0, #TODO: update FORs. For now assuming no outages.
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl:678]: If FOR data is avaialable, why don't we use it?

File: src/utils/reeds_data_parsing.jl [hunk 32]
```diff
@@ -480,8 +774,12 @@ end
         a Gen_Storage struct containing information about generators/storage
         technologies for each region.
 """
-function process_genstors(regions::Vector{<:AbstractString}, timesteps::Int)
-    gen_stors = Gen_Storage[] # empty for now
+function process_genstors(gen_stors, regions::Vector{<:AbstractString}, timesteps::Int)
+    if length(gen_stors)
```
Comment (scdhulipala) [inline src/utils/reeds_data_parsing.jl:805]: I'm not sure what's happening here.

File: src/utils/reeds_input_parsing.jl (hunk 27 above)
Comment (scdhulipala) [reply src/utils/reeds_input_parsing.jl:420]: If this is an acceptable implementation, I will implement this in https://github.com/NREL/ReEDS2PRAS/tree/dsc/general_cleanup

**Review by scdhulipala — CHANGES_REQUESTED** [review]

File: Project.toml (hunk 14 above)
Comment (sriharisundar) [reply Project.toml:4]: Sure.

File: src/utils/reeds_data_parsing.jl (hunk 19 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:260]: I was initially using this function directly for processing existing hydro plants as constant capacity. Can revert, or leave it as such.

File: src/utils/reeds_data_parsing.jl (hunk 21 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:323]: I initially did that, but felt this one should be used because we are modifying things in place, and I didn't want to change that code block

File: src/utils/reeds_data_parsing.jl (hunk 23 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:408]: Yes, the idea is to take all TODOs and create issues in ReEDS repo after merging this

File: src/utils/reeds_data_parsing.jl (hunk 25 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl]: Because that needs to be passed only if we want the energy limits to not be enforced.

File: src/utils/reeds_data_parsing.jl (hunk 26 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:532]: Since we specify returns above in the `if` block, I thought else is superfluous.

File: src/utils/reeds_data_parsing.jl (hunk 11 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:623]: No, we do not apply any derating at all

File: src/utils/reeds_data_parsing.jl (hunk 30 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:623]: No, we are not doing that for now. But can do so if necessary.

File: src/utils/reeds_data_parsing.jl (hunk 31 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:678]: Because we aren't disaggregating to individual units. That means if there is an outage, we lose all the HD capacity in that region, which can be on the order of GWs.

File: src/utils/reeds_data_parsing.jl (hunk 32 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:805]: This is checking if `gen_stors` has been populated earlier, and if so just returning it. If not, since we do not make `gen_stors` for any generators other than HD reservoirs, we just return empty.

File: src/utils/reeds_input_parsing.jl (hunk 27 above)
Comment (sriharisundar) [reply src/utils/reeds_input_parsing.jl:420]: Resolving this for now with the new commit from Surya

File: src/utils/reeds_data_parsing.jl (hunk 32 above)
Comment (scdhulipala) [reply src/utils/reeds_data_parsing.jl:805]: Okay. Do we need this functionm anymore after all the updates now?

File: src/utils/reeds_data_parsing.jl (hunk 29 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:571]: I am not sure what you're referring to here. Maybe we added lines and it moved.

File: src/utils/reeds_data_parsing.jl (hunk 25 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl]: Took care of this

File: src/utils/reeds_data_parsing.jl (hunk 16 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl]: Doing this only if we catch a `BoundsError` now. Added a TODO item so its taken care of in ReEDS.

File: src/utils/reeds_data_parsing.jl (hunk 32 above)
Comment (sriharisundar) [reply src/utils/reeds_data_parsing.jl:805]: I think we should keep it as a place holder to be sure we implement it if and when such data/devices become available.
